### PR TITLE
Improving generation

### LINKED ISSUE #2415: greedy beam search generates same sequence N times
## ❓ Questions & Help

I am doing greedy beam search (without sampling to avoid randomness) using GPT-2. However, all the returned sequences are same. Why is that the case? Shouldn't it give N best and different sequences?

```python
model = GPT2LMHeadModel.from_pretrained('gpt2-medium').cuda()
tokenizer = GPT2Tokenizer.from_pretrained('gpt2-medium')

input_context = 'The dog'
input_ids = torch.tensor(tokenizer.encode(input_context)).unsqueeze(0).cuda()
outputs = model.generate(input_ids=input_ids, num_beams=5, num_return_sequences=3)
for i in range(3):
    print('Generated {}: {}'.format(i, tokenizer.decode(outputs[0][i], skip_special_tokens=True)))
```

The resulting output is:
```
Generated 0: The dog was taken to a veterinary clinic for treatment.

The dog's owner said the!
Generated 1: The dog was taken to a veterinary clinic for treatment.

The dog's owner said the!
Generated 2: The dog was taken to a veterinary clinic for treatment.

The dog's owner said the!
````

### LINKED ISSUE #2482: model.generate should support past as an input
## 🚀 Feature

the `model.generate` method should support `past` as an input (and return the hidden states so that the next time it can inject past)

### LINKED ISSUE #2554: CTRL tokenizer has no special tokens to indicate EOS
## 🐛 Bug

The `generate` method from `PreTrainedModel` by default uses index 0 as EOS. This is a problem with CTRL, because its tokenizer has the word `the` mapped to this id.

Actually the CTRL has no special tokens besides UNK:

```
tokenizer = CTRLTokenizer.from_pretrained('ctrl')
tokenizer.special_tokens_map                                                                                                                                                                             
# {'unk_token': '<unk>'}

tokenizer.convert_ids_to_tokens([0])                                                                                                                                                                     
# ['the']
```

I believe the CTRL tokenizer should have some special token to use as EOS or PAD, as the other models do.

### PR DESCRIPTION
**Pull request by thomwolf**
Fix #2554 

TODO:
- add tests on generation

TODO potential:
- this PR could be used to fix #2415 and fix #2482 as well
- add TF 2.0 support for generation

### CODE REVIEW

**Comment by LysandreJik** [comment]
It would be great if this PR could handle the padding index for the models that do not have one. For example, GPT-2 doesn't have a padding index and therefore can't use the `generate` method, nor can it use the `batch_encode_plus` method.

**Comment by patrickvonplaten** [comment]
PR #2885 added the proposed changes.
